File: bidding-engine/src/main/java/com/omnibus/bidding/api/ApiErrorHandler.java
package com.omnibus.bidding.api;

import java.time.OffsetDateTime;
import java.util.Map;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.MethodArgumentNotValidException;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.bind.annotation.RestControllerAdvice;
import org.springframework.web.bind.support.WebExchangeBindException;

@RestControllerAdvice
public class ApiErrorHandler {

    @ExceptionHandler(IllegalArgumentException.class)
    public ResponseEntity<Map<String, Object>> handleIllegalArgument(IllegalArgumentException exception) {
        return build(HttpStatus.BAD_REQUEST, exception.getMessage());
    }

    @ExceptionHandler(MethodArgumentNotValidException.class)
    public ResponseEntity<Map<String, Object>> handleValidation(MethodArgumentNotValidException exception) {
        return build(HttpStatus.BAD_REQUEST, "Request validation failed");
    }

    @ExceptionHandler(WebExchangeBindException.class)
    public ResponseEntity<Map<String, Object>> handleWebFluxValidation(WebExchangeBindException exception) {
        return build(HttpStatus.BAD_REQUEST, "Request validation failed");
    }

    private ResponseEntity<Map<String, Object>> build(HttpStatus status, String message) {
        return ResponseEntity.status(status).body(Map.of(
            "timestamp", OffsetDateTime.now().toString(),
            "status", status.value(),
            "error", status.getReasonPhrase(),
            "message", message
        ));
    }
}

File: services/core/bidding-engine/src/main/kotlin/pl/writeonly/omnibus/bidding/api/ApiErrorHandler.kt
package pl.writeonly.omnibus.bidding.api

import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.MethodArgumentNotValidException
import org.springframework.web.bind.annotation.ExceptionHandler
import org.springframework.web.bind.annotation.RestControllerAdvice
import java.time.OffsetDateTime

@RestControllerAdvice
class ApiErrorHandler {
    @ExceptionHandler(IllegalArgumentException::class)
    fun handleIllegalArgument(exception: IllegalArgumentException): ResponseEntity<Map<String, Any?>> =
        build(HttpStatus.BAD_REQUEST, exception.message)

    @ExceptionHandler(MethodArgumentNotValidException::class)
    fun handleValidation(@Suppress("UNUSED_PARAMETER") exception: MethodArgumentNotValidException): ResponseEntity<Map<String, Any?>> =
        build(HttpStatus.BAD_REQUEST, "Request validation failed")

    private fun build(status: HttpStatus, message: String?): ResponseEntity<Map<String, Any?>> =
        ResponseEntity.status(status).body(
            mapOf(
                "timestamp" to OffsetDateTime.now().toString(),
                "status" to status.value(),
                "error" to status.reasonPhrase,
                "message" to message,
            ),
        )
}
